internal/builtinfunctions/functions_test.go

```diff
@@ -0,0 +1,471 @@
+package builtinfunctions_test
+
+import (
+	"go.flow.arcalot.io/engine/internal/builtinfunctions"
+	"math"
+	"reflect"
+	"testing"
+)
+
+var testData = map[string]struct {
+	functionName   string
+	rawArguments   []any
+	returnError    bool
+	expectedResult any
+}{
+	"int-to-float-positive": {
+		"intToFloat",
+		[]any{int64(1)},
+		false,
+		1.0,
+	},
+	"int-to-float-negative": {
+		"intToFloat",
+		[]any{int64(-1)},
+		false,
+		-1.0,
+	},
+	"int-to-float-zero": {
+		"intToFloat",
+		[]any{int64(0)},
+		false,
+		0.0,
+	},
+	"float-to-int-positive": {
+		"floatToInt",
+		[]any{1.0},
+		false,
+		int64(1),
+	},
+	"float-to-int-negative": {
+		"floatToInt",
+		[]any{-1.0},
+		false,
+		int64(-1),
+	},
+	"float-to-int-zero": {
+		"floatToInt",
+		[]any{0.0},
+		false,
+		int64(0),
+	},
+	"int-to-string-positive": {
+		"intToString",
+		[]any{int64(11)},
+		false,
+		"11",
+	},
+	"int-to-string-zero": {
+		"intToString",
+		[]any{int64(0)},
+		false,
+		"0",
+	},
+	"int-to-string-negative": {
+		"intToString",
+		[]any{int64(-11)},
+		false,
+		"-11",
+	},
+	"float-to-string-positive-whole": {
+		"floatToString",
+		[]any{11.0},
+		false,
+		"11",
+	},
+	"float-to-string-negative-whole": {
+		"floatToString",
+		[]any{-11.0},
+		false,
+		"-11",
+	},
+	"float-to-string-positive-fractional": {
+		"floatToString",
+		[]any{11.1},
+		false,
+		"11.1",
+	},
+	"float-to-string-negative-fractional": {
+		"floatToString",
+		[]any{-11.1},
+		false,
+		"-11.1",
+	},
+	"float-to-string-zero": {
+		"floatToString",
+		[]any{0.0},
+		false,
+		"0",
+	},
+	"float-to-string-nan": {
+		"floatToString",
+		[]any{math.NaN()},
+		false,
+		"NaN",
+	},
+	"bool-to-string-true": {
+		"boolToString",
+		[]any{true},
+		false,
+		"true",
+	},
+	"bool-to-string-false": {
+		"boolToString",
+		[]any{false},
+		false,
+		"false",
+	},
+	"string-to-int-positive": {
+		"stringToInt",
+		[]any{"11"},
+		false,
+		int64(11),
+	},
+	"string-to-int-negative": {
+		"stringToInt",
+		[]any{"-11"},
+		false,
+		int64(-11),
+	},
+	"string-to-int-zero": {
+		"stringToInt",
+		[]any{"0"},
+		false,
+		int64(0),
+	},
+	"string-to-int-invalid-1": {
+		"stringToInt",
+		[]any{"zero"},
+		true,
+		nil,
+	},
+	"string-to-int-invalid-2": {
+		"stringToInt",
+		[]any{"1.0"},
+		true,
+		nil,
+	},
```

Comment:
Correct. This needs to account for outputs from tools that they want to convert to a numeric literal.

Reply:
Why are the requirements for this function different from the requirements for literals in expressions?  Shouldn't they be consistent with each other?

Reply:
With expression literals, they're following a defined language. The restrictions need to make sense in the context of everything else in the language.
With the conversion functions, we're not limited by what makes sense as a literal in an expression. It could be used for converting a string output of a tool to an integer, for example. And we know that it's supposed to be an integer, rather than something else, because the function is `stringToInt`.

Reply:
I don't think they should be compared. They serve entirely different purposes under entirely different restrictions and contexts.